influxql/query_manager.go

```diff
@@ -49,8 +56,33 @@ type QueryParams struct {
 	// Not required, but highly recommended. If this channel is not set, the
 	// query needs to be manually managed by the caller.
 	InterruptCh <-chan struct{}
+
+	// Holds any error thrown by the QueryManager if there is a problem while
+	// executing the query. Optional.
+	Error *QueryError
+}
+
+// QueryError is an error thrown by the QueryManager when there is a problem
+// while executing the query.
+type QueryError struct {
+	err error
+	mu  sync.Mutex
 }
```

Comment:
This is something that can be done better by modifying the query executor to accept an error channel instead of a struct channel. The reason is because the mutex is surrounding the error and not the task since the way to get an error is to pass in this object to the query manager. If this object is present, it will be filled with an error.

I took this from how I've seen other libraries deal with optional metadata sections getting passed in. I know in the Consul Go client if you want metadata returned, you have to create the metadata struct and then call the function passing in that struct. It will then fill the struct you pass in rather than creating its own. This uses the same mechanism.

So in conclusion, it's because this mutex is protecting the error, not the task.